src/main/java/org/openrefine/extensions/commons/utils/CommonsImportingController.java

```diff
@@ -182,9 +183,124 @@ private void doParsePreview(
             job.updating = false;
         }
 
-    /*private static void parsePreview( ) {
-        
-    }*/
+    private static void parsePreview(
+            Project project,
+            ProjectMetadata metadata,
+            final ImportingJob job,
+            int limit,
+            ObjectNode options,
+            List<Exception> exceptions) throws IOException {
+
+        JSONUtilities.safePut(options, "headerLines", 0);
+        String cmtitle = JSONUtilities.getString(options, "docUrl", null);
+        if (!cmtitle.startsWith("Category:")) {
+            cmtitle = "Category:" + cmtitle;
+        }
+
+        parse(
+                project,
+                metadata,
+                job,
+                cmtitle,
+                limit,
+                options,
+                exceptions
+        );
+    }
+
+    static public void parse(
+            Project project,
+            ProjectMetadata metadata,
+            final ImportingJob job,
+            String cmtitle,
+            int limit,
+            ObjectNode options,
+            List<Exception> exceptions) throws IOException {
+
+            String pageName = cmtitle;
+            String fileSource = pageName;//FIXME: add filename
+
+            setProgress(job, fileSource, 0);
+
+            TabularImportingParserBase.readTable(
+                    project,
+                    job,
+                    new FilesBatchRowReader(job, fileSource, cmtitle),
+                    limit,
+                    options,
+                    exceptions);
+            setProgress(job, fileSource, 100);
+        }
+
+        static private void setProgress(ImportingJob job, String fileSource, int percent) {
+            job.setProgress(percent, "Reading " + fileSource);
+        }
+
+        static private class FilesBatchRowReader implements TableDataReader {
+            final ImportingJob job;
+            final String fileSource;
+            JsonNode files;
+            String cmtitle;
+            String cmcontinue;
+            String urlBase;
+            private int indexRow = 0;
+            List<Object> rowsOfCells;
+
+            public FilesBatchRowReader(ImportingJob job, String fileSource, String cmtitle) throws IOException {
+
+                this.job = job;
+                this.fileSource = fileSource;
+                this.cmtitle = cmtitle;
+                setURL();
+                getFiles(urlBase);
+
+            }
+            public void setURL() {
+
+                urlBase = "https://commons.wikimedia.org./w/api.php"
+                    + "?action=query&list=categorymembers&cmtitle="
+                    + cmtitle + "&cmtype=file&cmprop=title|type|ids&cmlimit=500&format=json";
```

Comment:
Here you are forming a URL by inserting a user-supplied value without URL-escaping it, which is potentially a vulnerability. Instead of using string concatenation to form the full URL, it would be better to only pass the bare URL `https://commons.wikimedia.org/w/api.php` to OkHTTP, and pass the GET parameters separately (I expect you can supply them in the form of a `Map<String, String>` or something similar). Then, OkHTTP will take care of the URL encoding for you, avoiding this problem.